### ISSUE TITLE
CSS refactor: Per-component styling

### ISSUE THREAD

**Issue by richardolsson** [zetkin/organize.zetk.in#129]
The SCSS source is starting to get big. As it gets bigger, the need for structure and consensus gets more pressing. This is a suggestion for how we could structure SCSS going forward.

### Per-component styling
Right now it's unnecessarily difficult to guess the class name, and even more so the SCSS source location, for a particular JSX component. This is completely unnecessary, since all components already have (more or less) unique names, with corresponding file names.

I propose that the top element of every component be given a `className` that is an all lowercase version of the component name, e.g. `PersonList` becomes `personlist`. Furthermore, the SCSS source pertaining to the `.personlist` selector is placed in `PersonList.scss` _next to_ the `PersonList.jsx` component JavaScript source code. For example, the file structure might look something like this:

* `src/components/personlist/`
  * `PersonList.jsx`
  * `PersonList.scss`
  * `PersonListItem.jsx`
  * `PersonListItem.scss`

The SCSS files could be collected automatically, e.g. using [gulp-sass-glob](https://www.npmjs.com/package/gulp-sass-glob).

### Responsive breakpoints
Media queries should go in each component SCSS file, so that all the styling related to a particular component can be found in one place, as a general rule of thumb. The actual pixel widths of breakpoints should be externalized to some central location, e.g. using variables. A SitePoint article describes [some solutions for this](http://www.sitepoint.com/managing-responsive-breakpoints-sass/).

### Common styles
Some styles and SCSS assets will be shared across the entire project. Those should still go in the `src/scss` folder, along with the `style.scss`.

### Caveats
There are some risks and possible pitfalls associated with this solution. 

#### Non-deterministic concatenation
If the SCSS files are collected automatically there will be no way of pre-determining the order in which they are concatenated. That means that order can't be relied upon for overrides. This means the only other option (ruling out `!important`) is specificity.

#### Name collisions 
Although most components have globally unique names, they are only required to be unique within a folder. This means that there could (and likely will) be occurrences of components with the same name appearing in two contexts. This means components can never have generic names like `Button` or `Footer` for the risk of causing conflicts.

## SCSS style
Below is an example of what the SCSS file for `MyComponent.jsx` would look like.

```scss
// MyComponent.scss
// Single top level block
.mycomponent {
  // Base, mobile-first styling
  width: 98%;
  margin: 0 1%;

  // Elements below top level have classnames consisting of the
  // component class name with a descriptive, locally unique suffix.
  .mycomponent-subelem {
    font-weight: 120%;
  }

  // The MyComponent.jsx file may contain styling for some other
  // component if that styling is local to this particular context, i.e.
  // only when OtherComponent appears within MyComponent.
  .othercomponent {
    background-color: black;
    color: white;
  }

  // Responsive breakpoints, using mixin
  @include respond-to(medium) {
    float: left;
    width: 48%;

    // Sub-element can re-occur here, with responsive styling
    .mycomponent-subelem {
      border-bottom: 1px solid #eee;
    }
  }
}
```

**Comment by richardolsson** [comment]
No rush, but please have a look at this idea for a SCSS structure when you have time, @j0n and @Bagera, and we can discuss it at some future hackathon.

After an initial discussion, I'll draft up a SCSS style guide that we can then create pull requests towards to discuss any specific issues. Right now I just want to know what you think of the general concept.

**Comment by richardolsson** [comment]
Because of #95, we will likely want to switch from Browserify to Webpack, so maybe this approach could work for CSS: http://bensmithett.com/smarter-css-builds-with-webpack/

**Comment by j0n** [comment]
Thinks the styleguide looks really great, I would just try to avoid some nestling. 

```
// MyComponent.scss
// Single top level block
.mycomponent {
  // Base, mobile-first styling
  width: 98%;
  margin: 0 1%;
}
// Skip first nestling, forces more unique names
// and there is less risk for nestling in nestling 
// and saving some bytes :)
.mycomponent-subelem {
  font-weight: 120%;
}
...
@include respond-to(medium) {
  float: left;
  width: 48%;
  .mycomponent-subelem {
    border-bottom: 1px solid #eee;
  }
}
```

states/modifiers should always handle by the jsx-code? And then theres is no need for `.mycomponent-subelem--modifier` in the styleguide? maybe a note about that?

**Comment by richardolsson** [comment]
I understand why you don't need to nest `.mycomponent-subelem`, but I wanted to do so anyway because it would make it consistent with when you have sub-components (not the same as elements) which do have their own styling somewhere else, but need to have specific styling for this context. They will need to be nested.

For example, you have a `Button` component, defined in `Button.jsx` and with it's default, global styling in `Button.css`. But when used in `MyComponent`, you want the button to have some degree of special styling, so you need to have `.button` nested within `.mycomponent`:

```scss
.mycomponent {
  .button {
    // Button instances in MyComponent should be red
    background-color: red;
  }
}
```

I'm not quite sure whether I think it's good or bad to have `.mycomponent-subelem` (which is _not_ it's own component) follow that same pattern or not. Perhaps not nesting in these cases is actually better, to further distinguish styling of sub-elements from context-specific styling of sub-components.

On a separate note however, I am not quite sure how the responsive breakpoint would work in your example. Would that not in fact result in a compile error? Because unless I'm mistaken, the `float: left` and `width: 48%` in this case are not actually in any element selector. Wouldn't your code generate the following, _invalid_ CSS?

```css
@media screen and (min-width: 720px) {
  // These are not enclosed in a selector block
  float: left;
  width: 48%;

  .mycomponent-subelem {
    border-bottom: 1px solid #eee;
  }
}
```

That's the reason why I wanted to nest the media query. To avoid having to add `.mycomponent` within the media query scope as well as in the file scope. But I guess you could add it to the media query, like so:

```scss
.mycomponent {
  // ...
}

.mycomponent-subelem {
  // ...
}

@include respond-to(medium) {
  .mycomponent {
    float: left;
    width: 48%;
  }

  .mycomponent-subelem {
    border-bottom: 1px solid #eee;
  }
}
```

Do you think that's better than my original suggestion?

**Comment by j0n** [comment]
Yes, you are right it would be an compile error with the media query, sloppy copy paste by me. 
Good you converted it to what I'm intended.

Yes, in my opinion its better

Regarding the nested components, there we need nested css. Hopefully it will be just a few lines of code and it reflects its a sub-component (`button`) inside the component (`mycomponent`). 
I would also put that styling outside the base style for the component (`mycomponent`), to make it clear its styling for sub-components.

So if I copy paste again... 
```scss
.mycomponent {
  // ...
}

.mycomponent-subelem {
  // ...
}

.mycomponent {
  .button {
    // Button instances in MyComponent should be red
    background-color: red;
  }
}

@include respond-to(medium) {
  .mycomponent {
    float: left;
    width: 48%;
  }

  .mycomponent-subelem {
    border-bottom: 1px solid #eee;
  }
}
```

**Comment by richardolsson** [comment]
Ok, I don't feel strongly either way. Do you prefer SASS nesting over explicit CSS descendant selectors?

```scss
// Nesting
.mycomponent {
  .button {
    // Button instances in MyComponent should be red
    background-color: red;
  }
}

// Descendant selector
.mycomponent .button {
  // Button instances in MyComponent should be red
  background-color: red;
}
```

I reckon the latter (descendant selectors) are fewer lines of code and if we're not gonna use any of the benefits of nesting (e.g. not having to repeat the ancestor class name for several descendants) I see no reason to use nesting at all in this case.

**Comment by j0n** [comment]
No, either way works for me

**Comment by richardolsson** [comment]
I have begun authoring a style guide in [a separate repository](https://github.com/zetkin/contributors-reference/tree/master/style/scss). This will be the place to collect all contributors documentation for the entire project going forward.

From now on, this issue will focus solely on getting the refactor for organize.zetk.in done according to the style guide. Any new ideas or objections to the style? Submit a pull request against the [contributors-reference](/zetkin/contributors-reference) repository.

**Comment by richardolsson** [comment]
The bulk of this (although not perfectly consistent) was implemented by PR #221.

**Comment by richardolsson** [comment]
Aside from needing some better documentation in the contributor's reference, this should be considered completed.
